### PR TITLE
Allow user to provide batch id on submitting batch job

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

This PR proposes to allow the user to provide a batch id on submitting a batch job. If the batch id already existed in metastore, Kyuubi ignores this submission and just returns the existing one, w/ a marker in response, this could avoid duplicated batch job submission.

Talking about the implementation, the key things are

How does the user set the custom batch id?

- User can optionally set the `kyuubi.batch.id` in `conf: Map[String, String]`, and the value must be a UUID, for Java users, it can be generated by `UUID.randomUUID().toString()`

How does the Kyuubi Server detect the duplication?

- It's simple in single Kyuubi Server instance case, Kyuubi just needs to look up the metastore before creating a batch job
- In HA mode, suppose the user requests to create the batch jobs w/ the same batch id concurrently, multiple Kyuubi Servers may process the request and try to insert to metastore DB at the same time, but only the first insertion success, others will fail w/ "duplicated key", Kyuubi Server needs to catch this error and return the existing batch job information instead of creating a new one.

How does the user know if the returned batch job is new created or duplicated?

- a new field `batchInfo: Map[String, String]` is added to the response, and for duplicated batch job, `"kyuubi.batch.duplicated": "true"` will be contained.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala [hunk 1]
```diff
@@ -210,22 +210,35 @@ private[v1] class BatchesResource extends ApiRequestContext with Logging {
...
-    val sessionHandle = sessionManager.openBatchSession(
-      userName,
-      "anonymous",
-      ipAddress,
-      request.getConf.asScala.toMap,
-      request)
-    buildBatch(sessionManager.getBatchSessionImpl(sessionHandle))
+    val userProvidedBatchId = request.getConf.asScala.get(KYUUBI_SESSION_BATCH_ID_KEY)
+      // the user provides batch_id must be an UUID
+      .map { batchId => UUID.fromString(batchId); batchId }
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala:215]: should be provide the more detail error message if the specified id is not an UUID.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala [hunk 2]
```diff
@@ -210,22 +210,35 @@ private[v1] class BatchesResource extends ApiRequestContext with Logging {
...
+
+    userProvidedBatchId.flatMap { batchId =>
+      Option(sessionManager.getBatchFromMetadataStore(batchId))
+    } match {
+      case Some(batch) =>
+        warn(s"duplicated submission: ${batch.getId}, ignore and return the existing batch.")
+        batch
+      case None =>
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala:223]: if the existing batch is None.

we need remove the `KYUUBI_SESSION_BATCH_ID_KEY` from batchRequest.getConf

**Review by turboFei — CHANGES_REQUESTED** [review]
left some comments

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala [hunk 3]
```diff
@@ -210,22 +210,35 @@ private[v1] class BatchesResource extends ApiRequestContext with Logging {
...
-        KYUUBI_SESSION_BATCH_RESOURCE_UPLOADED_KEY -> isResourceFromUpload.toString,
-        KYUUBI_SESSION_REAL_USER_KEY -> fe.getRealUser())).asJava)
-    val sessionHandle = sessionManager.openBatchSession(
-      userName,
-      "anonymous",
-      ipAddress,
-      request.getConf.asScala.toMap,
-      request)
-    buildBatch(sessionManager.getBatchSessionImpl(sessionHandle))
+    val userProvidedBatchId = request.getConf.asScala.get(KYUUBI_SESSION_BATCH_ID_KEY)
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala:213]: I recommend to new a java conf map, and use `map.remove()`.

**Comment by turboFei** [comment]
seems mistake, let me think it twice.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala (hunk 2 above)
Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala:223]: mistake.

If batchId is specified,
- if the batch is existing, return the batch.
- if the batch does exist, use the specified batchId


If no batchId specified, generate a new one

**Comment by ulysses-you** [comment]
How about add a `traceToken` in rest client ? Trino support that and Kyuubi server should return traceToken back to client. Then developer can deduplicate by itself.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala (hunk 1 above)
Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala]: updated

**Comment by pan3793** [comment]
> How about add a `traceToken` in rest client ? Trino support that and Kyuubi server should return traceToken back to client. Then developer can deduplicate by itself.

I think it makes things more complex, IMO let the user provide a custom batch id does not has many risks, we already restrict the batch id MUST be a UUID.

**Comment by pan3793** [comment]
cc @turboFei @ulysses-you it's ready for the next review.

**Review by ulysses-you — APPROVED** [review]

**Review by turboFei — APPROVED** [review]

**Comment by turboFei** [comment]
TODO: render the batchInfo in kyuubi-ctl

**Comment by pan3793** [comment]
> render the batchInfo in kyuubi-ctl

Tracking in https://github.com/apache/kyuubi/issues/4438

**Comment by pan3793** [comment]
Thanks, merged to master/1.7

**Pull request closed without merging**